### DISCUSSION TITLE
Allow "eliding" the typing of all generic arguments

### DISCUSSION THREAD

**Discussion by colejohnson66** [dotnet/csharplang#7596] [Language Ideas] [answered]
Generic argument inference is great, but it still has some sharp edges. Namely, the ability to elide specification of the _entire_ generic qualifier if (even just) one type is not in the parameter list.

For the current state of C#, say I have some function that takes `T` and gives back `T`. In that situation, I can elide the `<T>` argument on the function call. However, if I have a function that takes `TInput` and gives back `TOutput`, I'm _forced_ to qualify the entirety of `<TInput, TOutput>` on the call.

```csharp
// function contents are irrelevant
public T OneArity<T>(T input) => ...;
public TOutput TwoArity<TInput>(TInput input) => ...;

// can elide the one-arity call, but can't for the two-arity call
int one = OneArity(42); // or OneArity<int>(42)
int two = TwoArity<double, int>(42.0);
```

## Proposal

I propose the "sink" keyword `_` be allowed as a type specifier in generics. Using the "sink specifier" would instruct the compiler to infer the type, just as it does already, but for only a portion of the specifier. In such a situation, one would be able to write the following:

```
int one = OneArity(42); // or OneArity<_>(42)
int two = TwoArity<_, int>(42.0);
```

## Specification Change

The C# specification ([section 8.4.2](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/types#842-type-arguments)) would be changed as follows:

```diff
 type_argument_list
     : '<' type_arguments '>'
     ;

 type_arguments
     : type_argument (',' type_argument)*
     ;   

 type_argument
     : type
+    : '_'
     ;
```

## Backwards Compatibility

The only possible hiccup this proposal could cause is for users who have a type called `_`. However, users generally don't do that. Regardless, the remedy to a conflict would be to consider `_` as an elision if, and only if, no type named `_` exists in the current scope.

**Comment by 333fred** [comment, marked as answer]
Duplicate of https://github.com/dotnet/csharplang/issues/1349, which @TomatorCZ is actively working on specing in https://github.com/dotnet/csharplang/pull/7582.
